File: frontend/src/types/environmentTier.ts
export interface EnvironmentTierResponse {
  id: number;
  tenant_id: number;
  name: string;
  description: string | null;
  /** The standard tier this maps onto, or null for a tenant-specific one. */
  category: string | null;
  color: string | null;
  display_order: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  /**
   * B5 — per-tier idle override, in days. NULL means "use the tenant's
   * environment_lifecycle_policy.idle_threshold_days" — a legitimate state,
   * not a missing value. A form must render this BLANK when null, never
   * pre-filled with the tenant default: pre-filling it turns every save into
   * an explicit override nobody asked for, silently detaching the tier from
   * future tenant-default changes.
   */
  idle_threshold_days: number | null;
}

export interface EnvironmentTierCreate {
  name: string;
  description?: string | null;
  color?: string | null;
  display_order?: number;
  is_active?: boolean;
  idle_threshold_days?: number | null;
}

export interface EnvironmentTierUpdate {
  name?: string;
  description?: string | null;
  color?: string | null;
  display_order?: number;
  is_active?: boolean;
  /**
   * Explicitly OMITTED vs explicitly NULL both matter here (the backend
   * reads `model_fields_set`): omit the key to leave the override alone,
   * send `null` to clear it back to "use the tenant default". Always send
   * the key when this form field is part of the draft, or a clear silently
   * becomes a no-op.
   */
  idle_threshold_days?: number | null;
}

File: backend/app/api/v1/schemas/environment_tier.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class EnvironmentTierCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: Optional[str] = None
    color: Optional[str] = Field(default=None, max_length=7)
    display_order: int = 0
    is_active: bool = True
    # B5 — per-tier idle override. NULL (the default) means "use the
    # tenant's environment_lifecycle_policy.idle_threshold_days".
    idle_threshold_days: Optional[int] = Field(default=None, ge=1, le=3650)


class EnvironmentTierUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    color: Optional[str] = Field(default=None, max_length=7)
    display_order: Optional[int] = None
    is_active: Optional[bool] = None
    # Unlike the other nullable fields on this schema (description, color),
    # which the service only ever SETS (never clears — see
    # environment_tier_service.update_tier's `is not None` checks), an
    # explicit null here must be honoured: it is how an admin removes a
    # tier's override and falls back to the tenant default. The service reads
    # this via `model_fields_set`, the same pattern environment_service uses
    # for `expires_at`.
    idle_threshold_days: Optional[int] = Field(default=None, ge=1, le=3650)


class EnvironmentTierResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    name: str
    description: Optional[str] = None
    category: Optional[str] = None
    color: Optional[str] = None
    display_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
    idle_threshold_days: Optional[int] = None
